fix: account tracker controller with useMultiPolling

sahar-fehri opened the pull request:
## **Description**

Adds Multi chain polling for the token account tracker controller. 


We will have a separate PR to add the following e2e:
1- polling doesn't start with the wallet locked
2- once we unlock the wallet, polling starts with the wallet networks: we should check that the mocked requests for polling are happening
3- once we add a new account, polling is updated: we should check new requests including the new address happen

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/28277?quickstart=1)

## **Related issues**

Fixes:
Related: https://github.com/MetaMask/metamask-extension/pull/28402
## **Manual testing steps**

This should not result in any new visual changes/new errors

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

sahar-fehri committed:
fix: account tracker controller with useMultiPolling

bergeron commented on app/scripts/metamask-controller.js:
```diff
@@ -2597,7 +2596,6 @@ export default class MetamaskController extends EventEmitter {
   }
 
   stopNetworkRequests() {
```
We probably also want to call `this.accountTrackerController.stopAllPolling();` in `onClientClosed`, just to be sure no polling loops leak, similar to the other polling controllers.

sahar-fehri replied:
Makes sense! [9fc9319](https://github.com/MetaMask/metamask-extension/pull/28277/commits/9fc931961cb36fde7e8d6e375262c63fadf94767)

sahar-fehri committed:
fix: add stopAllPolling onClientClosed

sahar-fehri committed:
Merge branch 'develop' into feat/multi-chain-for-account-tracker-controller-v2

sahar-fehri committed:
fix: fix

sahar-fehri committed:
fix: fix lint

sahar-fehri committed:
fix: update privacy snapshot

sahar-fehri committed:
fix: lint

bergeron commented on ui/hooks/useAccountTrackerPolling.ts:
```diff
@@ -0,0 +1,25 @@
+const useAccountTrackerPolling = () => {
+  // Selectors to determine polling input
+  const networkConfigurations = useSelector(getNetworkConfigurationsByChainId);
+  const availableNetworkClientIds = Object.values(
+    networkConfigurations,
+  ).flatMap((networkConfiguration) =>
+    networkConfiguration.rpcEndpoints.map(
+      (rpcEndpoint) => rpcEndpoint.networkClientId,
+    ),
+  );
```
We don't need to poll across every rpc endpoint, but rather just the default endpoint for each chain.  So I think this would be something like

```suggestion
  const availableNetworkClientIds = Object.values(networkConfigurations).map(
    (networkConfiguration) =>
      networkConfiguration.rpcEndpoints[
        networkConfiguration.defaultRpcEndpointIndex
      ].networkClientId,
  );
```

sahar-fehri replied:
Sorry am i wrong or both suggestions have the same result?

bergeron replied:
If there are multiple rpc endpoints configured for a network, it looks like the old code would poll all of them.  Whereas the suggested code would only poll the default endpoint.

sahar-fehri committed:
fix: wip

sahar-fehri commented on privacy-snapshot.json:
```diff
@@ -29,6 +29,8 @@
   "github.com",
   "goerli.infura.io",
   "lattice.gridplus.io",
+  "linea-mainnet.infura.io",
```
Yes

bergeron replied:
Are these still needed after mocks have been added?

sahar-fehri replied:
good point actually i dnt think they are needed anymore; pushing now

sahar-fehri committed:
fix: wip

sahar-fehri committed:
Merge branch 'develop' into feat/multi-chain-for-account-tracker-controller-v2

darkwing committed:
Merge branch 'develop' into feat/multi-chain-for-account-tracker-controller-v2

sahar-fehri committed:
fix: fix onboarding privacy e2e test

seaona committed:
Multichain debug (#28402)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**
List of failing tests:

- Malicious Confirmation Signature - Bad Domain @no-mmi displays alert
for domain binding and confirms
- Malicious Confirmation Signature - Bad Domain @no-mmi initiates and
rejects from confirmation screen
- Malicious Confirmation Signature - Bad Domain @no-mmi initiates and
rejects from alert friction modal
- Malicious Confirmation Signature - Bad Domain @no-mmi displays alert
for domain binding and confirms
- Malicious Confirmation Signature - Bad Domain @no-mmi initiates and
rejects from confirmation screen
- Malicious Confirmation Signature - Bad Domain @no-mmi initiates and
rejects from alert friction modal
- smart transactions @no-mmi Completes a Swap
- Simple Send Security Alert - Blockaid @no-mmi should not show security
alerts for benign requests
- Simple Send Security Alert - Blockaid @no-mmi should show security
alerts for malicious requests
- MultiRpc: should select rpc from modal
- MetaMask onboarding @no-mmi doesn't make any network requests to
infura before create new wallet onboarding is completed
- MetaMask onboarding @no-mmi doesn't make any network requests to
infura before onboarding by import is completed
- MetaMask onboarding @no-mmi doesn't make any network requests to
infura before create new wallet onboarding is completed
- MetaMask onboarding @no-mmi doesn't make any network requests to
infura before onboarding by import is completed
- 

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/28402?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

Swaps


https://github.com/user-attachments/assets/656ba20a-bc48-4f57-84d4-8b11e62a25cf


ENS


https://github.com/user-attachments/assets/aa23bb34-aeb0-48e9-af17-c98a82367437



## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: sahar-fehri <sahar.fehri@consensys.net>

sahar-fehri committed:
fix: fix add feature flag

sahar-fehri committed:
Merge branch 'develop' into feat/multi-chain-for-account-tracker-controller-v2

sahar-fehri committed:
fix: fix flag name

seaona commented:
# QA Findings and Questions

## Requests before onboarding starts
I'm seeing RPC requests to the default networks as well as to the token price API before I started onboarding.
I think we should not send any request until the user has finished onboarding, as they might change the RPC provider they want to use.

https://github.com/user-attachments/assets/01e2e6c0-e5dc-4e42-af4f-c880d5ce95df

Side: we should capture this by our e2e, so I'll check what we are missing there in order for this to make the tests fail :thinking: - that can be a separate topic though

## Incorrect RPC provider used for polling, if I have a custom provider
If I change my RPC provider to a custom one (ie for Mainnet, I use Flashbots), I still see requests happening to Infura provider in Mainnet, triggered by network polling

https://github.com/user-attachments/assets/8142ff09-87b7-4311-a2e4-90126874d2c7



## RPC requests sent every ~20seconds to all the networks that I have in the wallet
Just some thoughts/questions around this:
- Are the continuous RPC requests (around every ~20seconds) expected?

https://github.com/user-attachments/assets/43a65517-2f2e-428c-9865-14d9abc51550

- Is Infura aware of the change in volume that would happen? This will increase the amount Infura requests (as well as the other providers, but Infura is the most used in most of the networks) so possibly stress their network/limits. Tied to that, have we considered how rate limits would affect to this feature?

darkwing commented:
@seaona We've been in several, several meetings with the Infura teams.  We're implementing measures, coming in another PR, to allow remote feature flag control over (1) polling intervals and (2) chain blacklists, so Infura will have a good degree of control over requests so that we don't take them down.

bergeron commented:
Polling during onboarding / locked fixed here:
https://github.com/MetaMask/metamask-extension/pull/28465

seaona commented:
> @seaona We've been in several, several meetings with the Infura teams. We're implementing measures, coming in another PR, to allow remote feature flag control over (1) polling intervals and (2) chain blacklists, so Infura will have a good degree of control over requests so that we don't take them down.

thank you v much for this explanation, David :pray:

sahar-fehri commented:
Hey @seaona ! im not able to repro 
`Incorrect RPC provider used for polling, if I have a custom provider`
I can see it polling against the new rpc; are you on latest for this branc?

https://github.com/user-attachments/assets/a373c3d8-1c42-4c9c-adbd-fa6b839d546e

seaona commented:
thank you all for your comments. Summing up:
- :yellow_circle: Requests before onboarding starts: I see the issue partially fixed with the last changes. I see no requests sent to any RPC, but I do see requests sent to `https://price.api.cx.metamask.io/v2/chains/1/spot-prices?tokenAddresses=0x0000000000000000000000000000000000000000&vsCurrency=ETH&includeMarketData=true 
` . It can be fixed on a separate PR if needed, given that the feature is under a flag. Up to you :) cc @bergeron 
- :green_circle: Incorrect RPC provider used for polling, if I have a custom provider: I see this working correctly. I didn't have the last changes, and after pulling them I see it working as expected. Thanks @sahar-fehri 
- :green_circle: RPC requests sent every ~20seconds to all the networks that I have in the wallet: this has been clarified by @darkwing 's comment above. Thanks David!



Video for point 1:


https://github.com/user-attachments/assets/2c76130a-f8f0-400c-acfe-10986a4f2c84

sahar-fehri committed:
Merge branch 'develop' into feat/multi-chain-for-account-tracker-controller-v2

seaona reviewed (dismissed):
looks good from the e2e and testing side :fire: awesome work!!

bergeron committed:
Merge branch 'develop' into feat/multi-chain-for-account-tracker-controller-v2

bergeron commented:
resolved merge conflicts

sahar-fehri committed:
Merge branch 'develop' into feat/multi-chain-for-account-tracker-controller-v2

sahar-fehri committed:
fix: fix smart tx mock

sahar-fehri committed:
Merge branch 'develop' into feat/multi-chain-for-account-tracker-controller-v2

bergeron reviewed (dismissed):
LGTM.  tested with and without PORTFOLIO_VIEW

sahar-fehri committed:
Merge branch 'develop' into feat/multi-chain-for-account-tracker-controller-v2

sahar-fehri committed:
fix: rm changes from privacy-snapshot
